### PR TITLE
Add an AsyncSpanProcessor that uses Disruptor.

### PR DESCRIPTION
**Pull request by bogdandrutu**

### CODE REVIEW

File: sdk_contrib/async_processor/build.gradle [hunk 1]
```diff
@@ -0,0 +1,11 @@
+description = 'OpenTelemetry SDK Contrib'
+
+dependencies {
+    api project(':opentelemetry-api'),
+            project(':opentelemetry-sdk')
```
Comment (songy23) [inline sdk_contrib/async_processor/build.gradle]: nit: indentation seems off

File: sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+ * A {@link SpanProcessor} implementation that uses {@code Disruptor} to execute all the hooks on an
+ * async thread.
+ */
+@ThreadSafe
+public final class DisruptorAsyncSpanProcessor implements SpanProcessor {
+  // Number of events that can be enqueued at any one time. If more than this are enqueued,
+  // then subsequent attempts to enqueue new entries will block.
+  private static final int DEFAULT_DISRUPTOR_BUFFER_SIZE = 8192;
+  // The default value of the Disruptor behavior, blocks when no space available.
+  private static final boolean DEFAULT_BLOCKING = true;
```
Comment (songy23) [inline sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java]: Per https://github.com/open-telemetry/opentelemetry-specification/pull/130 I think we want to make default non-blocking?

File: sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  public void shutdown() {
+    disruptorEventQueue.shutdown();
+  }
+
+  /**
+   * Returns a new Builder for {@link DisruptorAsyncSpanProcessor}.
+   *
+   * @param spanProcessorList the {@code List<SpanProcessor>} to where the Span's events are pushed.
+   * @return a new {@link DisruptorAsyncSpanProcessor}.
+   * @throws NullPointerException if the {@code spanExporter} is {@code null}.
```
Comment (songy23) [inline sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java]:
```suggestion
   * @throws NullPointerException if the {@code spanProcessorList} is {@code null}.
```

File: sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+     * @return this.
+     */
+    public Builder setWaitingStrategy(WaitStrategy waitingStrategy) {
+      this.waitStrategy = waitingStrategy;
+      return this;
+    }
+
+    /**
+     * Returns a new {@link DisruptorAsyncSpanProcessor} that converts spans to proto and forwards
+     * them to the given {@code spanExporter}.
```
Comment (songy23) [inline sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java]:
```suggestion
     * them to the given {@code spanProcessorList}.
```

File: sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+     *
+     * @return a new {@link DisruptorAsyncSpanProcessor}.
+     * @throws NullPointerException if the {@code spanExporter} is {@code null}.
```
Comment (songy23) [inline sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java]:
```suggestion
     * @throws NullPointerException if the {@code spanProcessorList} is {@code null}.
```

File: sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorEventQueue.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  void enqueue(ReadableSpan readableSpan, boolean isStart) {
+    if (isShutdown) {
+      if (!loggedShutdownMessage.getAndSet(true)) {
+        logger.log(Level.INFO, "Attempted to enqueue entry after Disruptor shutdown.");
+      }
+    }
+
+    if (blocking) {
+      long sequence = ringBuffer.next();
+      try {
```
Comment (songy23) [inline sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorEventQueue.java]: nit: this and https://github.com/open-telemetry/opentelemetry-java/pull/468/files#diff-6c02fc3d87a7b36b5f8badb3359dd315R107 can be combined.

**Comment by songy23** [comment]
Looks like there're test failures:
https://travis-ci.org/open-telemetry/opentelemetry-java/jobs/569063648#L682
```bash
io.opentelemetry.sdk.contrib.trace.export.DisruptorAsyncSpanProcessorTest > incrementMultiSpanProcessor FAILED
    expected: 1
    but was : 0
        at io.opentelemetry.sdk.contrib.trace.export.DisruptorAsyncSpanProcessorTest.incrementMultiSpanProcessor(DisruptorAsyncSpanProcessorTest.java:147)
io.opentelemetry.sdk.contrib.trace.export.DisruptorAsyncSpanProcessorTest > incrementOnce FAILED
    expected: 1
    but was : 0
        at io.opentelemetry.sdk.contrib.trace.export.DisruptorAsyncSpanProcessorTest.incrementOnce(DisruptorAsyncSpanProcessorTest.java:87)
5 tests completed, 2 failed
```

File: sdk/src/main/java/io/opentelemetry/sdk/trace/MultiSpanProcessor.java [hunk 7]
```diff
@@ -16,15 +16,29 @@
...
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
-final class MultiSpanProcessor implements SpanProcessor {
+/**
+ * Implementation of the {@code SpanProcessor} that simply forwards all received events to a list of
+ * {@code SpanProcessor}s.
+ */
+public final class MultiSpanProcessor implements SpanProcessor {
```
Comment (songy23) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/MultiSpanProcessor.java]: /cc @mayurkale22 this may be what you need.

Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/MultiSpanProcessor.java]: ACK.

**Comment by bogdandrutu** [comment]
@songy23 I fixed the tests and ensured that the async SpanProcessor waits for all the next SpanProcessors to finish.

File: sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java (hunk 2 above)
Comment (songy23) [reply sdk_contrib/async_processor/src/main/java/io/opentelemetry/sdk/contrib/trace/export/DisruptorAsyncSpanProcessor.java]: Now that open-telemetry/opentelemetry-specification#130 is merged this default value should be false (can be done in another PR).

**Pull request merged**